bdns/servers.go

```diff
@@ -0,0 +1,159 @@
+package bdns
+
+import (
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// serverProvider represents a type which can provide a list of addresses for
+// the bdns to use as DNS resolvers. Different implementations may provide
+// different strategies for providing addresses, and may provide different kinds
+// of addresses (e.g. host:port combos vs IP addresses).
+type ServerProvider interface {
+	Addrs() ([]string, error)
+	Stop()
+}
+
+// staticProvider stores a list of host:port combos, and provides that whole
+// list in randomized order when asked for addresses. This replicates the old
+// behavior of the bdns.impl's servers field.
+type staticProvider struct {
+	servers []string
+}
+
+var _ ServerProvider = &staticProvider{}
+
+func NewStaticProvider(servers []string) *staticProvider {
+	return &staticProvider{servers: servers}
+}
+
+func (sp *staticProvider) Addrs() ([]string, error) {
+	if sp.servers == nil || len(sp.servers) == 0 {
```

Comment:
Note that `len(nil) == 0`, so `len(sp.servers) == 0` would cover both of these cases.

Reply:
I _know_ this but it always makes me feel uncomfortable for some reason, so I always hedge. Thanks for the reminder.